### PR TITLE
MAINT: Check for `__array_ufunc__` before doing anything else.

### PR DESCRIPTION
**Pull request by mhvk**
This helps avoid unnecessary cleanup.

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -4952,6 +4947,7 @@ ufunc_at(PyUFuncObject *ufunc, PyObject *args)
         return override;
     }
 
+    NPY_BEGIN_THREADS_DEF;
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This is a definition, so has to be before any statements (yay, pre-C99...)

Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Oops - now corrected. Possibly makes more sense if the real implementation is in a new function, but I'll leave that for another PR (I'm again looking at moving the implementation of ndarray ufunc to `ndarray.__array_ufunc__`, or at least making the call there faster).

**Comment by eric-wieser** [comment]
Doesn't look like you pushed the fix

**Comment by mhvk** [comment]
Indeed, silly of me. Now done.

File: numpy/core/src/umath/ufunc_object.c [hunk 2]
```diff
@@ -4078,26 +4078,22 @@ ufunc_generic_call(PyUFuncObject *ufunc, PyObject *args, PyObject *kwds)
...
-
     errval = PyUFunc_CheckOverride(ufunc, "__call__", args, kwds, &override);
     if (errval) {
         return NULL;
     }
     else if (override) {
-        for (i = 0; i < ufunc->nargs; i++) {
-            PyArray_DiscardWritebackIfCopy(mps[i]);
-            Py_XDECREF(mps[i]);
-        }
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Wait, why was this even here? There's no way things can be non-null...

**Comment by eric-wieser** [comment]
Thanks!

File: numpy/core/src/umath/ufunc_object.c (hunk 2 above)
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: That's right - that's what originally triggered me, and why I removed it. It was part of a much larger PR (#9639) and I think just an oversight.

**Pull request merged**